src/server/game/Entities/Object/Object.cpp

```diff
@@ -1884,23 +1884,25 @@ TempSummon* Map::SummonCreature(uint32 entry, Position const& pos, SummonPropert
     if (summoner)
         phase = summoner->GetPhaseMask();
 
+    Unit* summonerUnit = summoner ? summoner->ToUnit() : nullptr;
+
     TempSummon* summon = nullptr;
     switch (mask)
     {
         case UNIT_MASK_SUMMON:
             summon = new TempSummon(properties, summoner, false);
             break;
         case UNIT_MASK_GUARDIAN:
-            summon = new Guardian(properties, summoner, false);
+            summon = new Guardian(properties, summonerUnit, false);
```

Comment:
Such a shame that `Minion` and other derived classes rely on `Unit* Minion::GetOwner()` never being null. And that they have this `Minion::GetOwner()` in the first place, which "conveniently" shadows `Unit::GetOwner()`, no doubt causing a few bugs along the way.

There are at least 2 cases of trap-GOs summoning non-wild creatures in 3.3.5: spells 53076 and 54355. Very curious to see sniffs with what's happening there, but in the context of this PR it's not a big deal, since currently I imagine these spells don't work at all because of a guard condition in the beginning of `Spell::SummonGuardian` that would prevent non-unit spellcasters from summoning any guardians.

Reply:
It felt like I should not have added GameObject owner support to TempSummon types that I wasn't sure should have (like is a GameObject supposed to be a totem ?) but in the initial local changes I did I actually changed that too.
So if you have a couple of spells that are supposed to spawn a Guardian with a GameObject owner, it's quite easy to add support to that too.

Reply:
I think it might be easier to force such spells to summon "wild" TempSummons instead of Guardians if the caster isn't a Unit, until it can be proven that they should be Guardians. Client might not react well to it either, what would it even display, "Volatile Trap Knockback Bunny <Volatile Trap's Guardian>"? Would client be able to correctly inherit owner trap's faction for reaction checks? I doubt it. It's not worth the hassle to complicate the code further just for two spells which we aren't even sure how they worked on retail; for all we know, blizzards server could gleefully discard "ALLY/GUARDIAN" summon category if the spell is cast by a GO. And furthermore, they shouldn't currently be working in TC anyway (unless someone bothered to hackfix them) ever since the "GO casting" commit which happened years ago. And if no one raised a stink by now - they won't after this PR either, and it will get fixed when someone will get around to it.

Reply:
Found a sniff with 53079
<details>

```
ServerToClient: SMSG_UPDATE_OBJECT (0x0C22) Length: 778 ConnIdx: 0 Time: 11/16/2013 13:28:07.000 Number: 1032041
Map: 571
Count: 3
[0] UpdateType: DestroyObjects
[0] Object Count: 1

[2] UpdateType: CreateObject1
[2] GUID: Full: 0xF13072D5004A2348 Type: Creature Entry: 29397 (Land Mine Bunny) Low: 4858696
[2] Object Type: 3 (Unit)
[2] Has Attacking Target: False
[2] Transport Frames Count: 0
[2] Has Stationary Position: False
[2] Is Living: True
[2] Has Anim Kits: False
[2] Has Vehicle Data: False
[2] Has GameObject Position: False
[2] Has Scene Object Data: False
[2] Has GameObject Rotation: False
[2] bitA4: False
[2] Movement Flags: 1536 (OnTransport, DisableGravity)
[2] bit8C: True
[2] bit8D: False
[2] Spline Id: 234583155
[2] FloatBC: 7
[2] FloatAC: 8
[2] FloatA8: 2.5
[2] Time: 1629028945
[2] FloatB0: 4.5
[2] FloatC0: 4.5
[2] FloatC4: 3.141593
[2] FloatC8: 3.141593
[2] FloatB4: 4.72222
[2] FloatB8: 2.5
[2] GUID1: Full: 0xF13072D5004A2348 Type: Creature Entry: 29397 (Land Mine Bunny) Low: 4858696
[2] Position: X: 6116.151 Y: -655.6545 Z: 383.1613
[2] Orientation: 2.076939
[2] OBJECT_FIELD_GUID: 4858696/6.808483E-39
[2] 1: 4046484181/-8.73731E+29
[2] OBJECT_FIELD_TYPE: 9/1.261169E-44
[2] OBJECT_FIELD_ENTRY: 29397/4.119397E-41
[2] OBJECT_FIELD_SCALE_X: 1065353216/1
[2] UNIT_FIELD_BYTES_0: 33554688/9.404242E-38
[2] UNIT_FIELD_DISPLAY_POWER: 1/1.401298E-45
[2] UNIT_FIELD_HEALTH: 11770/1.649328E-41
[2] UNIT_FIELD_MAXHEALTH: 11770/1.649328E-41
[2] UNIT_FIELD_LEVEL: 78/1.093013E-43
[2] UNIT_FIELD_FACTIONTEMPLATE: 14/1.961818E-44
[2] UNIT_FIELD_FLAGS: 33554432/9.403955E-38
[2] UNIT_FIELD_FLAGS_2: 2048/2.869859E-42
[2] UNIT_FIELD_BASEATTACKTIME: 2000/2.802597E-42
[2] UNIT_FIELD_BASEATTACKTIME + 1: 2000/2.802597E-42
[2] UNIT_FIELD_BOUNDINGRADIUS: 1056964608/0.5
[2] UNIT_FIELD_COMBATREACH: 1065353216/1
[2] UNIT_FIELD_DISPLAYID: 11686/1.637557E-41
[2] UNIT_FIELD_NATIVEDISPLAYID: 11686/1.637557E-41
[2] UNIT_FIELD_MOD_CASTING_SPEED: 1065353216/1
[2] UNIT_FIELD_MOD_SPELL_HASTE: 1065353216/1
[2] UNIT_FIELD_MOD_HASTE: 1065353216/1
[2] UNIT_FIELD_MOD_RANGED_HASTE: 1065353216/1
[2] UNIT_FIELD_MOD_HASTE_REGEN: 1065353216/1
[2] UNIT_CREATED_BY_SPELL: 54355/7.616758E-41
[2] UNIT_FIELD_BYTES_2: 1/1.401298E-45
[2] UNIT_FIELD_HOVERHEIGHT: 1065353216/1
```
</details>

The spell is supposed to summon a Guardian, but as we can see here, summoned creatures have no owner GUID among their update fields. They do have the same level and faction as the summoner GO, but we don't have enough information to conclude that they inherit this data from summoner GO, because in CreatureDifficulty this creature 29397 also has level 78 and faction 14 set.

Reply:
Anyway, my suggestion would be to change Spell::EffectSummonType like this:
```diff
     switch (properties->Category)
     {
         case SUMMON_CATEGORY_WILD:
         case SUMMON_CATEGORY_ALLY:
         case SUMMON_CATEGORY_UNK:
             if (properties->Flags & 512)
             {
                 SummonGuardian(effIndex, entry, properties, numSummons);
                 break;
             }
-            switch (properties->Type)
+            switch (casterUnit ? properties->Type : SUMMON_TYPE_NONE)
             {
                case SUMMON_TYPE_PET:
                case SUMMON_TYPE_GUARDIAN:
                case SUMMON_TYPE_MINION:
                case SUMMON_TYPE_RUNEBLADE:
                case SUMMON_TYPE_CONSTRUCT:
                case SUMMON_TYPE_LIGHTWELL:
                case SUMMON_TYPE_BUTLER:
                    SummonGuardian(effIndex, entry, properties, numSummons);
                    break;
```
and a similar change to the code above the line that this comment is on. This would force GOs to summon wild tempsummons no matter what oddities might be in spell data. Unless Flags&512, that is... but there are no cases of this in 3.3.5, so I dunno.